Dashboard filters

agilliland opened the pull request:
Resolves #2556 

TODO:

frontend:
- [x] fade out cards that don't have an active parameter wired up
- [x] apply default parameter values when saving dashboard
- [x] get correct filter icon from Maz; replace `Parameters` tooltip text with `Add a filter`
- [x] give parameters a default name
- [x] red outline around parameter name input if duplicate
- [x] resolve conflict between duplicate cards (no longer safe assumption that same card = same data)
- [x] improved widget for selecting dimension/parameter mapping (dropdowns should only show dimensions that match the selected filter type)
  - [x] foreign keys
  - [x] SQL parameters
- [x] color code mapping widget button border based on validity
- [x] support SQL parameter mapping
- date pickers
  - [x] single date
  - [x] date range
  - [x] mm-yyyy
  - [x] q-yyyy
  - [x] relative
- [x] text widget
- [x] category picker w/ field-value
- [x] update URL when changing parameter values
- [x] resolve race condition on dashboard load
- [x] investigate why page scrolls when opening default value popover
- [x] ensure parameter mapping header modal thing always covers nav bar
- [x] strip out unused parameter mappings when saving dashboards (e.x. if a parameter or series is removed)
- [x] fix issue with card data not loading when first adding a card
- [x] filter parameter options by type
- [x] no way to un-map a card. putting the X icon on the right of the mapping dropdown if a selection has been made.
- [x] updated Add Filter icon, which has a plus sign incorporated to make it clearer that the action is about adding a filter, not applying a filter.
- [x] hover state of items in the Add Filter popup
- [x] try floating filters left
- [x] slimmer horizontal designs for filters
- [x] The "no overlap" warning on the mapping dropdowns is happening even when no selections have been made.
- [x] 0.5s CSS transition on cards fading in and out
- [x] “No Results” state for cards instead of displaying “There was a problem displaying this chart” warning.
- ~~It's kind of a pain that text input box filters are case-sensitive, but I know the team has discussed this before and felt that there are situations where case really matters. I just still think that if I type "san francisco" and nothing shows up, and then I type "San Francisco" and there are results, that feels dumb.~~ (this isn't currently implemented in the query processor)
- ~~I'm sometimes seeing a scenario where I select a filter widget, and map fields to it that should be low enough cardinality (like City), but then the resulting widget is an input box instead of a dropdown.~~ (this is due to those fields not having field values, not a frontend issue) 
- [x] mapping dropdowns for multi-series cards spacing is off
- [x] mapping dropdowns for multi-series cards need labels on top of them to clarify which series each dropdown belongs to
- [x] Info message on multi-series card to inform users that they have to map ALL series in the card to a filter in order for the card to work with it at all
- [x] Dropdowns that have no valid options in them should be in a disabled state with text that says, “No valid [type of widget, e.g. date] fields." I'm seeing this in some cases instead.
- [x] I can’t get month+year filters to work. See for example the Multi Series Demo dashboard and try using the month and year filter on it.
- [x] When editing the label of a filter, it should allow you to hit enter/return to commit change, and Esc to cancel it
- [x] When you select a value from a filter/parameter dropdown, it should commit the selection and close the dropdown (right now it’s not closing).
- [x] Loading indicator when a filter is being applied and the query is being re-run
- [x] Placeholder text for input-box filters where you can put in a default value should be something clearer like, “Enter an optional default value”
- [x] Hide Add Series or Remove Card actions when in filter-editing mode
- [x] Showing "No Results" instead of error message on dashboards in some cases
- [x] The clear filter X button for input field filters needs `cursor: pointer` applied
- [ ] Maybe an enhancement for later if we can't get to it, but we should probably require that filter labels not be blank — if it's blank and the user hits Enter, we could not commit the change, and turn the input box border red.
- [x] the add a filter icon still isn’t blue on hover as it should be
- [x] when you click on a date using the single date widget, it should close the popover
- [x] with the single date picker, if I open the popover when a value is already selected, and then click the blue X to clear the filter, I think that should also close the popover. I think this applies to all of the filter widgets, actually.
- [x] I should be seeing a heading above State that is whatever the FK field is in Orders that points to People, because the State field is actually in the People table. screen shot 2016-06-10 at 3 08 28 pm
- [x] fix issue with fk parameter options
- [x] fullscreen parameters treatment

backend:
- [x] more relative dates (`lastweek, lastmonth, lastyear, thisweek, thismonth, thisyear`)
- [x] quarter + year
- [x] month + year

agilliland committed:
very quick implementation of parameterized cards.

agilliland committed:
quick implementation of dashboard filters using parameterized cards.

agilliland committed:
quick and dirty application of url params to card params.

agilliland committed:
basics for parameterized SQL + some styling updates.

agilliland committed:
start breaking out parameter pickers into different types [wip].

tlrobinson committed:
Merge branch 'flow' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Tell UnusedFilesWebpackPlugin to ignore 'types' directories

agilliland committed:
more work on QP handling of parameters to flush out date range handling.

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

agilliland committed:
Merge branch 'master' into dashboard-filters

agilliland committed:
api and db support for saving :parameters on Dashboard.

agilliland committed:
fix up migration which was throwing an error.

tlrobinson committed:
Initial dashboard filter adding/editing UI

agilliland committed:
handling of :parameter_mappings attribute on DashboardCard

agilliland committed:
fix up database migration so that we get a proper NOT NULL column with the default value we want.  also some unit test fixes.

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Initial DashCardParameterMapping

tlrobinson committed:
Basic parameter mapping working end to end. Needs better dimension/parameter option logic

tlrobinson committed:
Crude ability to set parameter values, send those parameters with dataset query

agilliland committed:
Merge branch 'master' into dashboard-filters

agilliland committed:
updating structure of parameters to use `:target` attribute with specific structure for native parameters vs MBQL dimensions.

tlrobinson committed:
Improved parameter editing flow

tlrobinson committed:
Dashboard filters working end-to-end!

agilliland committed:
only do parameter substitution on :native cards if the driver supports :native-parameters

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Move Dashboard to main React router / Redux store, add initial URL param and default values

tlrobinson committed:
Cleanup text parameter input widget

tlrobinson committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Ensure parameter mapping works even when parameters == null

tlrobinson committed:
Cleanup package.json scripts

tlrobinson committed:
s/cardData/dashcardData/ and key off of dashcard.id+card.id to avoid conflicts in parameterized dashboard

tlrobinson committed:
Fade out cards with unmapped parameters. Ensure default parameters are used when saving. Ensure dashboard is reloaded when saved

tlrobinson committed:
Don't blur cards when editing

tlrobinson committed:
Set default paramter name, show red border when duplicated

tlrobinson committed:
Better filter button + tooltip

tlrobinson committed:
Non overlapping field value detection and field value widget

tlrobinson committed:
Ensure field values are loaded up front

tlrobinson committed:
single date picker

agilliland committed:
we will expect a parameter :type of "date/*" for the various date widgets.

tlrobinson committed:
Implement all parameter widgets

tlrobinson committed:
Mapping options for fk dimensions + SQL parameters

tlrobinson committed:
Update URL when parameter changes. Use URL-safe ~ as date separator

agilliland committed:
support for more date ranges options.

mazameli commented:
1. There seems to be no way to un-map a card. I overlooked this, and we’ll need a solution for this, like putting the X icon on the right of the mapping dropdown if a selection has been made.
2. I need to provide the updated Add Filter icon, which has a plus sign incorporated to make it clearer that the action is about adding a filter, not applying a filter.
3. I need to provide a design for the hover state of items in the Add Filter popup
4. I think the filter widgets are getting lost on the right. Can we try floating them left?
5. We need to try applying the horizontal filter labels and slimmer designs that I made in response to Sameer's comments. I'll tighten up the design that I think works best and supply that.
6. The "no overlap" warning on the mapping dropdowns is happening even when no selections have been made.
7. It would be nice if cards that get faded back when a filter is applied that they're not mapped to had a CSS transition of like 0.5s, both for fade out and fade in
8. We need a “No Results” state for cards instead of displaying “There was a problem displaying this chart” warning.
9. It's kind of a pain that text input box filters are case-sensitive, but I know the team has discussed this before and felt that there are situations where case really matters. I just still think that if I type "san francisco" and nothing shows up, and then I type "San Francisco" and there are results, that feels dumb.
10. I'm sometimes seeing a scenario where I select a filter widget, and map fields to it that should be low enough cardinality (like City), but then the resulting widget is an input box instead of a dropdown. 
11. We need to massage the mapping dropdowns for multi-series cards. The spacing is off, and they probably need labels on top of them to clarify which series each dropdown belongs to. Also, the designs show an Info message on multi-series card to inform users that they have to map ALL series in the card to a filter in order for the card to work with it at all.
12. Dropdowns that have no valid options in them should be in a disabled state with text that says, “No valid [type of widget, e.g. date] fields." I'm seeing this in some cases instead:

![screen shot 2016-06-07 at 11 31 08 am](https://cloud.githubusercontent.com/assets/2223916/15871554/2a2e5606-2caa-11e6-8945-098d872e0852.png)

13. I can’t get month+year filters to work. See for example the Multi Series Demo dashboard and try using the month and year filter on it.
14. When editing the label of a filter, it should allow you to hit enter/return to commit change, and Esc to cancel it. It feels weird right now.
15. When you select a value from a filter/parameter dropdown, it should commit the selection and close the dropdown (right now it’s not closing).
16. We probably need a loading indicator when a filter is being applied and the query is being re-run on a card so you know something’s happening. For heavy cards with multi-series, etc., it can be hard to tell that the query is being run. We could just use our standard card loading spinner.
17. The placeholder text for input-box filters where you can put in a default value isn’t clear. It’s currently, “Enter a value,” but we should change it to something clearer like, “Enter an optional default value”
18. I don’t think the Add Series or Remove Card actions should appear on cards when you’re in filter-editing mode.

agilliland committed:
update clj-time to 0.12.0, add a bunch more unit tests for parameters date range calculationsa along with a couple fixes.

camsaul commented on src/metabase/util.clj:
```diff
@@ -325,6 +325,13 @@
+  (or (nil? v)
+      (and (sequential? v)
+           (every? map? v))))
+
```
The name of this function is a bit misleading, since vectors and arrays are both `sequential?` but neither one is a `List`.

I would suggest breaking this up into two functions, a more general `maybe?` function that checks for `nil` or satisfaction of a predicate:


```clojure
(defn maybe? [pred x]
  (or (nil? x)
      (pred x)))

(maybe? string? nil)   ; -> true
(maybe? string? "ABC") ; -> true
(maybve? string? 1)    ; -> false
```

And a more general function for checking whether something is a *sequence* of maps:

```clojure
(defn sequence-of-maps? [x]
  (and (sequential? x)
       (every? map? x)))

(sequence-of-maps? nil)               ; -> false
(sequence-of-maps? [{:a :b} {:c :d}]) ; -> true
```

`sequence-of-maps?` is a more accurate name. The two can then be used generally or combined to get the desired effect:

```clojure
(maybe? sequence-of-maps? nil)        ; -> true
(maybe? sequence-of-maps? '({:a :b})) ; -> true
(maybe? sequence-of-maps? [1 2 3])    ; -> false
```

camsaul commented on src/metabase/driver/bigquery.clj:
```diff
@@ -472,8 +472,9 @@
           ;; TODO - either write BigQuery-speciifc tests for FK functionality or add additional code to manually set up these FK relationships for FK tables
-          :features              (constantly (when-not config/is-test?
-                                               #{:foreign-keys}))
+          :features              (constantly (merge #{:native-parameters}
```
This should be added to the list of valid features in the docstring for `metabase.driver/features`.

camsaul commented on src/metabase/models/dashboard.clj:
```diff
@@ -18,17 +19,53 @@
+       (events/publish-event :dashboard-create)))
+
+(defn update-dashboard
+  "Update a `Dashboard`"
```
These docstrings need some work. What's the purpose of this function? Why should I use it vs. just calling `db/update!`? Can I still use `db/update!` to update a `Dashboard`, or should I always use `update-dashboard`? What does `update-dashboard` do differently from `db/update!`?

Also don't forget functions that have side effects like these need to end in a `!`.

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,228 @@
+                               (s/replace incomplete-inner-clause "")))]
+    (update-in query-dict [:native :query] find-and-replace)))
+
+(defn- conjoin-multi-clause [clause]
```
It never hurts to write a quick blurb about the purpose of these functions. They might make sense to you, but I have no idea what they're for or what they're supposed to do.

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,228 @@
+(defn- absolute-date->range
+  "Take a given string description of an absolute date range and return a MAP with a given `:start` and `:end`.
+   Supported formats:
+      \"2014-05-10~2014-05-16\"
```
For markdown I think there needs to be a blank line before a code block

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,228 @@
+
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           PUBLIC API                                             |
```
If we're going to use boxes like these to separate sections of code we might as well take an extra second to do it right. The `|` at the end of the second line is misaligned here.

camsaul commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -100,12 +107,14 @@
                       :row          nil
                       :updated_at   true
                       :created_at   true
+                      :parameter_mappings []
```
indentation

mazameli commented:
Re: item 1

Here's what I'm thinking for how to clear out a field mapping (notice the green X):

![screen shot 2016-06-07 at 1 49 21 pm](https://cloud.githubusercontent.com/assets/2223916/15874151/c0844a1e-2cb6-11e6-982b-e0c59b65c4d0.png)

mazameli commented:
Re: item 11 in my comment above:

Here's the info message about multi-series cards when mapping fields to a filter. This also shows how i think the dropdowns should be spaced, and how we can put labels above the dropdowns to indicate which series they map to.

CSS for the background box:
```background: #F5F5F5;
border-radius: 2px;```

CSS for the text inside:
```font-family: Lato-Heavy;
font-size: 14px;
color: #8691AC;```

![screen shot 2016-06-07 at 1 48 45 pm](https://cloud.githubusercontent.com/assets/2223916/15874258/3197fda4-2cb7-11e6-8dba-d50930da3f83.png)

tlrobinson commented:
> It's kind of a pain that text input box filters are case-sensitive, but I know the team has discussed this before and felt that there are situations where case really matters. I just still think that if I type "san francisco" and nothing shows up, and then I type "San Francisco" and there are results, that feels dumb.

The same limitation exists in normal filters, right? Not sure if defaulting to case insensitive is "correct", but maybe we could make it configurable. Probably worth filing a separate issue.

> I'm sometimes seeing a scenario where I select a filter widget, and map fields to it that should be low enough cardinality (like City), but then the resulting widget is an input box instead of a dropdown.

Does the equivalent filter directly on a card show a dropdown or input box?

> I can’t get month+year filters to work. See for example the Multi Series Demo dashboard and try using the month and year filter on it.

I think @agilliland has fixed but not deployed this

mazameli commented:
Re: item 2 in my comments:

Here's the path for the filter icon with the plus symbol in it:
`M22.5185184,5.27947653 L17.2510286,5.27947653 L17.2510286,9.50305775 L22.5185184,9.50305775 L22.5185184,14.7825343 L26.7325102,14.7825343 L26.7325102,9.50305775 L32,9.50305775 L32,5.27947653 L26.7325102,5.27947653 L26.7325102,0 L22.5185184,0 L22.5185184,5.27947653 Z M14.9369872,0.791920724 C14.9369872,0.791920724 2.77552871,0.83493892 1.86648164,0.83493892 C0.957434558,0.83493892 0.45215388,1.50534608 0.284450368,1.77831828 C0.116746855,2.05129048 -0.317642562,2.91298361 0.398382661,3.9688628 C1.11440788,5.024742 9.74577378,17.8573356 9.74577378,17.8573356 C9.74577378,17.8573356 9.74577394,28.8183645 9.74577378,29.6867194 C9.74577362,30.5550744 9.83306175,31.1834301 10.7557323,31.6997692 C11.6784029,32.2161084 12.4343349,31.9564284 12.7764933,31.7333621 C13.1186517,31.5102958 19.6904355,27.7639669 20.095528,27.4682772 C20.5006204,27.1725875 20.7969652,26.5522071 20.7969651,25.7441659 C20.7969649,24.9361247 20.7969651,18.2224765 20.7969651,18.2224765 L21.6163131,16.9859755 L18.152048,15.0670739 C18.152048,15.0670739 17.3822517,16.199685 17.2562629,16.4000338 C17.1302741,16.6003826 16.8393552,16.9992676 16.8393551,17.7062886 C16.8393549,18.4133095 16.8393551,24.9049733 16.8393551,24.9049733 L13.7519708,26.8089871 C13.7519708,26.8089871 13.7318369,18.3502323 13.7318367,17.820601 C13.7318366,17.2909696 13.8484216,16.6759061 13.2410236,15.87149 C12.6336257,15.0670739 5.59381579,4.76288686 5.59381579,4.76288686 L14.9359238,4.76288686 L14.9369872,0.791920724 Z`

camsaul commented:
We've discussed making the string comparison filters case-insensitive before (#2370). I think it's a pretty good idea (definitely more user-friendly) and should be technically feasible; it will require a little bit of work since different DBs handle lowercasing strings in different ways.

On that issue though there were some arguments against making the filters case-insensitive (I think @agilliland was against it?)

mazameli commented:
>> It's kind of a pain that text input box filters are case-sensitive, but I know the team has discussed this before and felt that there are situations where case really matters. I just still think that if I type "san francisco" and nothing shows up, and then I type "San Francisco" and there are results, that feels dumb.
> The same limitation exists in normal filters, right? Not sure if defaulting to case insensitive is "correct", but maybe we could make it configurable. Probably worth filing a separate issue.

Yeah, this is the normal filter behavior, and we have discussed it in #2370 before, as @camsaul mentioned. Let's not get bogged down in this right now; we can revisit #2370 separately instead.

>> I'm sometimes seeing a scenario where I select a filter widget, and map fields to it that should be low enough cardinality (like City), but then the resulting widget is an input box instead of a dropdown.
> Does the equivalent filter directly on a card show a dropdown or input box?

Good call. I just checked, and it's an input box, so it must just be our logic for determining if a filter should be a dropdown or an input box is just different from what I was expecting.

>> I can’t get month+year filters to work. See for example the Multi Series Demo dashboard and try using the month and year filter on it.
> I think @agilliland has fixed but not deployed this
👌

tlrobinson commented:
That logic is just part of the backend sync process. Basically if a field has more than a certain number of distinct values (or you disable "in-depth database analysis") then it won't compute and store the field values, thus won't show the dropdown.

mazameli commented:
Re: item 3 in my comment above

Here's a hover state for the filter selection menu. It's just our normal style:

![screen shot 2016-06-07 at 2 57 18 pm](https://cloud.githubusercontent.com/assets/2223916/15876107/3acf2c36-2cc0-11e6-8de7-4483e8efd6d6.png)

mazameli commented:
Re: number 5 in my comment above

Here's what I think could work for side labels. The text for the labels in this mock is 13px. Note that the filters are placed on a white background with a divider above them. One obvious limitation of this layout is that we might run out of horizontal space. We could either A) set the filters to overflow:hidden, B) add a horizontal scroll if we need it, or C) and some kind of arrow buttons to scroll through the filters. I think we should start with option A and see how limiting it feels.

![side labels](https://cloud.githubusercontent.com/assets/2223916/15878871/87041e64-2cd3-11e6-8911-dc00305fa89f.png)

mazameli commented:
Re: number 8 above

@kdoh did you do a design somewhere for questions/cards with no results? I found this image on my desktop, but I don't know where it came from.

![dashboard-no-results](https://cloud.githubusercontent.com/assets/2223916/15879239/3f5f14e8-2cd7-11e6-9485-d7950378c687.png)

tlrobinson committed:
ensure parameter mapping header modal thing always covers nav bar

tlrobinson committed:
Fix issue with ActionButton reseting due to porting to ES6 class

tlrobinson committed:
Fix AddSeriesModal dashcardData loading

tlrobinson committed:
Fix dashboard database metadata loading

tlrobinson committed:
Remove unused parameter mappings when saving dashcards

tlrobinson committed:
Rework card data loading + misc other

tlrobinson committed:
Add a way to clear paramter mapping

tlrobinson committed:
Change funnel icon to version with plus, remove circle treatment, add hoverstate to add filter popovers

tlrobinson committed:
Green text in parameter mapper

tlrobinson committed:
Show titles and warning when mapping multiseries

tlrobinson committed:
Show spinner when refreshing, fade card in and out

tlrobinson committed:
Don't show spinner when auto-refreshing

tlrobinson committed:
Add no_results message to dashboards

tlrobinson committed:
Side by side filters

agilliland committed:
undo some of the QB parameters prototype work that we decided not to use.

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

agilliland committed:
formatting fix.

agilliland committed:
updated documentation and naming to clarify things a bit more.

agilliland committed:
couple more updates on style/naming.

agilliland committed:
fix migration.

agilliland committed:
Merge branch 'master' into dashboard-filters

tlrobinson committed:
Parameter option type filtering, improved widget and popover

tlrobinson committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Ensure category popover closes when set

tlrobinson committed:
Better placeholder text when editing

agilliland committed:
go back to clj-time 0.11.0

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Continue to show edit/remove buttons when editing another parameter, but fade them out

tlrobinson committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

agilliland committed:
fix bigquery :features so that we don't get nils

agilliland committed:
fix reflection warnings.

mazameli commented:
Things are looking good! Some small things I'm seeing:

- tiny detail: the `add a filter` icon still isn’t blue on hover as it should be
- when you click on a date using the single date widget, it should close the popover
- with the single date picker, if I open the popover when a value is already selected, and then click the blue X to clear the filter, I think that should also close the popover. I think this applies to all of the filter widgets, actually.
- One funny edge case: the `quantified cat` litter box table (I love my job) includes a field called `state`, which the location filter mistook as being a geographical state, and allowed me to map it to the State filter

tlrobinson commented:
Things are looking good! Some small things I'm seeing:

> One funny edge case: the `quantified cat` litter box table (I love my job) includes a field called `state`, which the location filter mistook as being a geographical state, and allowed me to map it to the State filter

Nice catch, but this is unrelated to dashboard filters. It looks like we mark any text fields with the name `state` to be `special_type: "state"`. This seems like it may be a moderately common edge case though, so it may be worth filing a separate issue.

https://github.com/metabase/metabase/blob/master/src/metabase/models/field.clj#L180

> In this scenario, I'm trying to map a State (Location) filter to this `Top Orderers` card, which is based on the `Orders` table. I'm able to see a `State` field in this list, but according to the conversation we had with Sameer yesterday, I should be seeing a heading above `State` that is whatever the FK field is in `Orders` that points to `People`, because the `State` field is actually in the `People` table.
<img width="792" alt="screen shot 2016-06-10 at 3 08 28 pm" src="https://cloud.githubusercontent.com/assets/2223916/15980118/79e9c188-2f1d-11e6-8c3c-0a7ae739b1e0.png">

I thought we decided if there were no FK options we shouldn't show the original table's heading? Doesn't matter to me, I can easily change it if that's what we want.

camsaul commented on src/metabase/driver/bigquery.clj:
```diff
@@ -473,8 +473,8 @@
-          :features              (constantly (when-not config/is-test?
-                                               #{:foreign-keys}))
+          :features              (constantly (cond-> #{:native-parameters}
+                                                     (not config/is-test?) :foreign-keys))
```
what are you trying to accomplish here? This code doesn't make sense and will break foreign key support for BigQuery. e.g. 

```clojure
(cond-> #{:native-parameters}
  true :foreign-keys) 
;; -> nil

(cond-> #{:native-parameters}
  false :foreign-keys)
;; -> #{:native-parameters}
```

```

:-1:

mazameli commented:
> I thought we decided if there were no FK options we shouldn't show the original table's heading? Doesn't matter to me, I can easily change it if that's what we want.

We did, but that's not the scenario here: the card is based on the Orders table, but this State field is not in that table; it's in the People table.

camsaul commented on src/metabase/query_processor.clj:
```diff
@@ -501,22 +512,24 @@
+              post-check-results-format
+              pre-log-query
+              guard-multiple-calls
+              run-query) (assoc query :driver driver))))))
```
looks like you had a bad merge here. Adding `pre-substitute-parameters` should just be a single line, not reverting this whole section to what the code was a few weeks ago 👎

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           MBQL QUERIES                                             |
```
indentation

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           SQL QUERIES                                             |
+;;; +-------------------------------------------------------------------------------------------------------+
```
indentation

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+       (assoc param :target [target (str param-name ":end")],   :value end)])))
+
+(defn- substitute-param [param-name value query]
+  ;; TODO: escaping and protection against SQL injection!
```
+1

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+(def ^:private ^:const outer-clause-prefix-pattern #"^\[\[(.*?)\s.*\]\]$")
+(def ^:private ^:const incomplete-outer-clause-pattern #"\[\[.*?\{\{.*?\}\}.*?\]\]")
+(def ^:private ^:const inner-clause-pattern #"<(.*?)>")
+(def ^:private ^:const incomplete-inner-clause-pattern #"<.*?\{\{.*?\}\}.*?>")
```
It would be nice to see these constants [indented together](https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#align-single-line-vars)

And it would be extremely [useful](https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#judiciously-use-comments) to see some quick blurbs about the purpose of these functions. I or any potential contributors won't have any idea what they're for or what they're supposed to do.

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+                                               (if outer-clause
+                                                 (recur (s/replace-first sql outer-clause (format-outer-clause outer-clause)) rest)
+                                                 sql))))
+    query-dict))
```
Can you just use `clojure.walk` here instead?

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           PUBLIC API                                             |
+;;; +-------------------------------------------------------------------------------------------------------+
```
indentation

camsaul commented:
Are there any tests here that run a query with parameters against an actual database? 😟

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,248 @@
+                                             (loop [sql                      q
+                                                    [[orig stripped] & rest] inner-clauses]
+                                               (if orig
+                                                 (recur (s/replace-first sql orig stripped) rest)
```
why can't you just replace all the inner clauses at once with `s/replace` instead of doing them one-at-a-time and recursing?

camsaul commented on test/metabase/query_processor/parameters_test.clj:
```diff
@@ -0,0 +1,314 @@
+  (expand-parameters {:database   1
+                      :type       :query
+                      :query      {:aggregation ["rows"]
+                                   :breakout    [17]}
```
field references need to be written like `[field-id 17]`, not `17`

camsaul commented on test/metabase/query_processor/parameters_test.clj:
```diff
@@ -0,0 +1,314 @@
+;; TODO: test for value escaping to prevent sql injection (native queries)
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           NATIVE QUERIES                                             |
```
indentation

tlrobinson committed:
Fix getDimensions fk-> support

tlrobinson committed:
Improve dashboard 'no results' message logic

tlrobinson committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

tlrobinson committed:
Close SingleDate widget when clickign a date

tlrobinson committed:
ParameterWidget 'X' should have cursor-pointer and close popover when clicked

tlrobinson committed:
Add filter icon blue on hover

mazameli commented:
One thing we overlooked: how filters and the "No results" illustration look in fullscreen night mode.

![screen shot 2016-06-10 at 5 07 32 pm](https://cloud.githubusercontent.com/assets/2223916/15981685/fe9fe82a-2f2d-11e6-88e1-a1385db71e42.png)

tlrobinson committed:
Show mapping option section headers if there are any fk options

tlrobinson committed:
Hide action buttons when editing parameter mappings

tlrobinson committed:
Fullscreen parameters treatment

tlrobinson committed:
Lint fixes

agilliland committed:
comments and formatting.

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

agilliland committed:
disable native parameter substitution until we are going to use it.

tlrobinson committed:
Fix adding to dashboard from qb

tlrobinson committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

mazameli commented:
Seeing a funny popover menu placement bug on the second choice for the Add Filter dialogue. It's most noticeable when you select a Location widget, but it's also happening when you select a Time widget (notice the popover's top arrow).

![screen shot 2016-06-13 at 2 06 43 pm](https://cloud.githubusercontent.com/assets/2223916/16023176/68b5f5e4-3170-11e6-8b83-f299f23ff5af.png)

mazameli commented:
It would be super cool 😎  if we could get the spacing above dashboard filter dropdowns/inputs to be the same as the spacing below them. Current state:

![spacing](https://cloud.githubusercontent.com/assets/2223916/16023302/11020d64-3171-11e6-8e01-7280ca43ab7f.png)

mazameli commented:
It seems funny that cards that aren't wired up to a particular filter still show the query loading animation when a value is selected for that filter. I think, ideally, they'd just fade back as soon as a filter they're not wired up to gets applied. #enhancement-for-later-probably

agilliland committed:
additional tests.

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

camsaul commented on test/metabase/query_processor_test.clj:
```diff
@@ -16,7 +16,8 @@
             [metabase.test.util :as tu]
-            [metabase.util :as u]))
+            [metabase.util :as u]
+            [metabase.query-processor :as qp]))
```
`:require` ordering 😥

camsaul commented on frontend/src/metabase/dashboard/components/Dashboard.jsx:
```diff
@@ -65,19 +65,20 @@ export default class Dashboard extends Component {
-            await this.props.fetchDashboard(this.props.selectedDashboard);
-            if (this.props.addCardOnLoad) {
+            await fetchDashboard(this.props.selectedDashboard);
+            if (addCardOnLoad != null) {
```
Should this be `!==` instead?

tlrobinson replied:
This is equivalent to `addCardOnLoad !== null && addCardOnLoad !== undefined`. It's the one place it's acceptable to use `!=` in JavaScript.

camsaul replied:
Ok, that makes sense, thanks 👍

mazameli commented:
Something weird's going on with this field dropdown. Looks like we need a max height, but there's also something funny going on with the lack of field icons. This is the `damage type` card on the `Product Stuff` dashboard being mapped to a Category filter.

![screen shot 2016-06-13 at 2 47 09 pm](https://cloud.githubusercontent.com/assets/2223916/16024344/3cdec8c8-3176-11e6-9529-9b2b9d0ab504.png)

camsaul commented on frontend/src/metabase/dashboard/components/parameters/widgets/DateRelativeWidget.jsx:
```diff
@@ -0,0 +1,59 @@
+    "thisweek":   "This Week",
+    "thismonth":  "This Month",
+    "thisyear":   "This Year",
+};
```
Would it make sense to instead structure these like:

```javascript
const FILTERS = {
    today: {
        name: "Today",
        mapping: ["=", null, ["relative_datetime", "current"]]
    },
    yesterday: {...}, 
    ...}
```

(A single map of maps instead of two separate maps?)

That way it'd be easier to keep track of the different filters and add new ones if we want (e.g. something like "This week")

mazameli commented:
I think we looked at this last week, @tlrobinson, but I'm still seeing `Cannot read property "id" of undefined` when I try adding a card to a dashboard on proto.

camsaul commented on frontend/src/metabase/lib/utils.js:
```diff
@@ -1,5 +1,10 @@
+    return Math.floor((1 + Math.random()) * 0x10000)
+      .toString(16)
+      .substring(1);
+}
```
`s4` is a pretty confusing function name 😢 . How about something more descriptive like `rand4CharHexStr`?

camsaul commented:
In the data migration, what is advantage of adding the `NOT NULL` constraint separately?

e.g. instead of doing
```yaml
changes:
  - addColumn:
      tableName: report_dashboard
      columns:
        - column:
            name: parameters
            type: text
            constraints:
              nullable: true
              deferrable: false
              initiallyDeferred: false
  - addNotNullConstraint:
      columnDataType: text
      columnName: parameters
      defaultNullValue: '[]'
      tableName: report_dashboard
  - addColumn:
      tableName: report_dashboardcard
      columns:
        - column:
            name: parameter_mappings
            type: text
            constraints:
              nullable: true
              deferrable: false
              initiallyDeferred: false
  - addNotNullConstraint:
      columnDataType: text
      columnName: parameter_mappings
      defaultNullValue: '[]'
      tableName: report_dashboardcard
```

instead [you can just do](http://www.liquibase.org/documentation/column.html):

```yaml
changes:
  - addColumn:
      tableName: report_dashboard
      columns:
        - column:
            name: parameters
            type: text
            defaultValue: '[]'
            constraints:
              nullable: false
              deferrable: false
              initiallyDeferred: false
  - addColumn:
      tableName: report_dashboardcard
      columns:
        - column:
            name: parameter_mappings
            type: text
            defaultValue: '[]'
            constraints:
              nullable: false
              deferrable: false
              initiallyDeferred: false
```

which is clearer and a lot more concise

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,252 @@
+    "Q1" (t/first-day-of-the-month (.withMonthOfYear (t/date-time year) DateTimeConstants/JANUARY))
+    "Q2" (t/first-day-of-the-month (.withMonthOfYear (t/date-time year) DateTimeConstants/APRIL))
+    "Q3" (t/first-day-of-the-month (.withMonthOfYear (t/date-time year) DateTimeConstants/JULY))
+    "Q4" (t/first-day-of-the-month (.withMonthOfYear (t/date-time year) DateTimeConstants/OCTOBER))))
```
This would be a lot more concise if you made the `condp` the last argument of `.withMonthOfYear`. e.g.

```clojure
(defn- start-of-quarter [quarter year]
  (t/first-day-of-the-month (.withMonthOfYear (t/date-time year) (condp = quarter
                                                                   "Q1" DateTimeConstants/JANUARY
                                                                   "Q2" DateTimeConstants/APRIL
                                                                   "Q3" DateTimeConstants/JULY
                                                                   "Q4" DateTimeConstants/OCTOBER))))
```

camsaul commented:
I've mentioned this previously on other PRs but you need to use `yyyy-MM-dd` in date format strings, not `YYYY-MM-dd`. [`YYYY` is the "week year"](http://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html), which means the year the current week started. So Jan 1st 2016 becomes "2015-01-01" unless the first was a Sunday.

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,252 @@
+      (let [{:keys [start end]} (absolute-date->range param-value)]
+        ["BETWEEN" field start end])
+      ;; relative date range
+      (condp = param-value
```
You don't need to use `condp =` here. Just use `case`.

```clojure
(case param-value
  "past7days" ["TIME_INTERVAL" field -7 "day"]
  ...)
```

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,252 @@
+  (let [tz        (t/time-zone-for-id report-timezone)
+        formatter (tf/formatter "YYYY-MM-dd" tz)
+        today     (.withTimeAtStartOfDay (t/to-time-zone (t/now) tz))]
+    (->> (condp = value
```
same thing. You should use `case` here instead of `condp =`

tlrobinson commented:
@camsaul I assume you're only referring to the usage of `YYYY` on the backend? It looks like moment.js' `YYYY` is actually what we want on the frontend. Fun. https://tonicdev.com/tlrobinson/5756047e8b908f1300210f4e

camsaul commented on src/metabase/query_processor/parameters.clj:
```diff
@@ -0,0 +1,252 @@
+      (substitute-all-params query-dict rest)
+      (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
+        (substitute-all-params query rest)))
+    query-dict))
```
this would probably be clearer written as a `cond`.

```clojure
(defn- substitute-all-params [query-dict [{:keys [value], [_ param-name] :target, :as param} & rest]]
  (cond
    (not param)               query-dict
    (and (string? param-name)
         value)               (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
                                (substitute-all-params query rest))
    :else                     (substitute-all-params query-dict rest)))
```

At any rate `(and param param-name value (string? param-name))` can be simplified to 

```clojure
(and value (string? param-name))
```

since you already checked `param` in the line above and `(string? param-name)` will fail on `nil`

camsaul commented:
@tlrobinson Yeah I was referring to the backend only. `DD` also means "day of year" as far as Java is concerned so it's important we capitalize the format string exactly as `yyyy-MM-dd` on the backend.  I guess moment must be a little less cray-cray

camsaul commented on src/metabase/models/dashboard.clj:
```diff
@@ -18,17 +19,53 @@
 
+(defn- pre-insert [dashboard]
+  (let [defaults {:parameters []}]
+    (merge defaults dashboard)))
```
If `[]` is the column-level default value of `:parameters` we shouldn't need this step. It would be a little faster to let the DB take care of it

tlrobinson committed:
Params spacing

tlrobinson committed:
Combine relative date filters metadata

tlrobinson committed:
Ensure popover position updates when size changes

tlrobinson committed:
Scroll options list, add default icon

tlrobinson commented:
Ok I think all the known frontend issues should be fixed.

@agilliland Are you handling the backend changes or should I?

mazameli commented:
Here's that animation enhancement for the instructions if we can squeak it in.

[move down animation.zip](https://github.com/metabase/metabase/files/312924/move.down.animation.zip)

agilliland committed:
few more tweaks and fixes.

agilliland commented:
@tlrobinson I made a bunch of the updates to the backend, so everything should be g2g

tlrobinson committed:
Fix dashboard loading/switching

camsaul commented:
There's still a bunch of points about the backend code I've brought up over the past few weeks that haven't been so much as acknowledged. 

-  [ ] Name of `nil-or-maplist?` is misleading since argument it returns `true` even when things aren't lists
-  [ ] `nil-or-maplist?` would be much more useful if broken into two separate functions
-  [ ] `create-dashboard` / `update-dashboard` should be called `create-dashboard!` / `update-dashboard!` ([this is a standard Clojure convention for functions w/ side-effects](functions that have side effects like these need to end in a !))
-  [ ] Not really clear if `update-dashboard` is supposed to be the only way one should update a dashboard in the future, or if `db/update!` is acceptable as well; this ought to be clarified in the docstring
-  [ ] Some changes in `query_processor.clj` accidentally reverted some changes I made last week
-  [ ] Indentation is off in a bunch of places
-  [ ] Standard Clojure convention is to use `case`, not `condp =` (several places)
-  [ ] `pre-insert` probably isn't necessary in `dashboard.clj`
-  [ ] DB migration could be substantially simplified
-  [ ] There might be some opportunities for  substantial code simplification by using `walk` instead of `loop`

At the very least the merit a discussion

agilliland commented:
I changed the name of `nil-or-maplist?` quite a while ago.

I believe the naming, docstrings, and implementation of `create-dashboard` and `update-dashboard` are fine.  I have already discussed with you a couple times the fact that I do not agree with your interpretation of the use of the `!` suffix in the case of our database functions, so I do not consider that a standard that has ever been agreed upon.

I have fixed all of the code indentation issues that I saw.

The rest of your remarks are potential other ways of writing code, but they have no functional impact so I consider them optional.  There are undoubtably hundreds of other ways that any of this code could be written, but if the current code looks fine and is working then I intend to stick with it.

camsaul commented:
The [link I referenced regarding function naming conventions](https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark) is from the standard Clojure style guide. It's a universal convention in Clojure. You can ask a hundred other Clojure developers and they'll all tell you the same thing: functions that have unsafe side-effects need to end in a `!`.

RE your other point: Yes, code can be written in different ways. But some ways are clearer, more concise, or more flexible than others. It's silly to finish 99% of a PR and refuse to listen to any suggestions about ways to improve the code because the current code is working. 

👎

salsakran commented:
@camsaul the prioritization framework for reviewing code should be (in order)

* calling out a need to write tests 
* pointing out user facing bugs
* pointing out corner cases where incorrect results might occur
* pointing out performance issues
* pointing out approaches that close off features we expect to work on in the next 6-12 months
and LASTLY deviations from coding style.

Let's focus energy on what matters.

tlrobinson committed:
Header modal animation

tlrobinson committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

camsaul commented:
@salsakran I've done a mix of all those things, @agilliland listened to my suggestions about things that would break existing features, missing tests, etc. I agree that style issues are lower-priority, but they're not zero-priority.

camsaul commented:
And if suggestions for improving code are offered, they deserve a fair consideration.

agilliland committed:
appease the linter.

agilliland committed:
fix up unit tests.

agilliland committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

camsaul committed:
Fix bad merge :unamused:

camsaul committed:
Fix indentation :unamused:

camsaul committed:
Merge branch 'dashboard-filters' of github.com:metabase/metabase into dashboard-filters

camsaul committed:
Use `case` instead of `condp =` (idiomatic Clojure) :unamused:

camsaul committed:
minor code simplifcation :yum:

camsaul commented:
Does it make sense to leave a bunch of commented out code in this PR? It would probably make sense just to move those to a separate WIP PR
